core/internal/runhistoryreader/parquet/iterator/parquet_data_iterator.go

```diff
@@ -9,64 +9,74 @@ import (
 // ParquetDataIterator iterates over history steps in a run's parquet file.
 type ParquetDataIterator struct {
 	ctx             context.Context
-	selectedRows    *SelectedRows
+	selectedRows    *SelectedRowsRange
 	selectedColumns *SelectedColumns
+	fileReader      *pqarrow.FileReader
 
 	recordReader pqarrow.RecordReader
 	current      *RowGroupIterator
+
+	lastStep int64
 }
 
 func NewParquetDataIterator(
 	ctx context.Context,
 	reader *pqarrow.FileReader,
-	selectedRows *SelectedRows,
+	selectedRows *SelectedRowsRange,
 	selectedColumns *SelectedColumns,
-) (RowIterator, error) {
-	columnIndices := selectedColumns.GetColumnIndices()
-	rowGroupIndices, err := selectedRows.GetRowGroupIndices()
-	if err != nil {
-		return nil, err
+) (*ParquetDataIterator, error) {
+	pqDataIterator := &ParquetDataIterator{
+		ctx:             ctx,
+		selectedRows:    selectedRows,
+		selectedColumns: selectedColumns,
+		fileReader:      reader,
+		lastStep:        0,
 	}
-
-	recordReader, err := reader.GetRecordReader(
-		ctx,
-		columnIndices,
-		rowGroupIndices,
-	)
+	err := pqDataIterator.createNewRecordReader()
 	if err != nil {
 		return nil, err
 	}
+	return pqDataIterator, nil
+}
 
-	// No rows to read
-	if !recordReader.Next() {
-		return &NoopRowIterator{}, nil
-	}
+func (r *ParquetDataIterator) GetSelectedRows() *SelectedRowsRange {
+	return r.selectedRows
+}
 
-	current, err := NewRowGroupIterator(
-		recordReader.RecordBatch(),
-		selectedColumns,
-		selectedRows,
+func (r *ParquetDataIterator) UpdateQueryRange(
```

Comment:
needs a doc explaining conditions under which this can be used, and the post-conditions after an error (is this still guaranteed to be a valid ParquetDataIterator? Does it need to be Released immediately?)